src/hotspot/os/linux/cgroupV1Subsystem_linux.hpp

```diff
@@ -103,31 +119,28 @@ class CgroupV1Subsystem: public CgroupSubsystem {
     const char * container_type() {
       return "cgroupv1";
     }
-    CachingCgroupController * memory_controller() { return _memory; }
-    CachingCgroupController * cpu_controller() { return _cpu; }
+    CachingCgroupController<CgroupMemoryController>* memory_controller() { return _memory; }
+    CachingCgroupController<CgroupCpuController>* cpu_controller() { return _cpu; }
 
   private:
     /* controllers */
-    CachingCgroupController* _memory = nullptr;
+    CachingCgroupController<CgroupMemoryController>* _memory = nullptr;
     CgroupV1Controller* _cpuset = nullptr;
-    CachingCgroupController* _cpu = nullptr;
+    CachingCgroupController<CgroupCpuController>* _cpu = nullptr;
     CgroupV1Controller* _cpuacct = nullptr;
     CgroupV1Controller* _pids = nullptr;
 
-    jlong read_mem_swappiness();
-    jlong read_mem_swap();
-
   public:
     CgroupV1Subsystem(CgroupV1Controller* cpuset,
-                      CgroupV1Controller* cpu,
+                      CgroupV1CpuController* cpu,
                       CgroupV1Controller* cpuacct,
                       CgroupV1Controller* pids,
-                      CgroupV1MemoryController* memory) {
-      _cpuset = cpuset;
-      _cpu = new CachingCgroupController(cpu);
-      _cpuacct = cpuacct;
-      _pids = pids;
-      _memory = new CachingCgroupController(memory);
+                      CgroupV1MemoryController* memory) :
+      _memory(new CachingCgroupController<CgroupMemoryController>(memory)),
+      _cpuset(cpuset),
+      _cpu(new CachingCgroupController<CgroupCpuController>(cpu)),
+      _cpuacct(cpuacct),
+      _pids(pids) {
     }
```

Comment:
Pre-existing: No destructor freeing these.

Reply:
A `CgroupSubsystem` instance lives as long as the JVM lives. I'm not sure this is worth the extra complexity. For example, I'd rather avoid copying those path strings around and free them again (`CgroupV1Controller` and `CgroupV2Controller` destructors don't free, for example, because they get re-used latter avoiding this extra copying around). Instead essentially use those singletons and let the OS cleanup handle de-allocation.